src/main/java/com/commercetools/sync/services/impl/TypeServiceImpl.java

```diff
@@ -33,13 +53,62 @@ public TypeServiceImpl(@Nonnull final BaseSyncOptions syncOptions) {
     @Override
     public CompletionStage<Optional<String>> fetchCachedTypeId(@Nonnull final String key) {
         if (!isCached) {
-            return fetchAndCache(key);
+            return cacheAndFetch(key);
         }
         return CompletableFuture.completedFuture(Optional.ofNullable(keyToIdCache.get(key)));
     }
 
     @Nonnull
-    private CompletionStage<Optional<String>> fetchAndCache(@Nonnull final String key) {
+    @Override
+    public CompletionStage<Set<Type>> fetchMatchingTypesByKeys(@Nonnull final Set<String> keys) {
+        if (keys.isEmpty()) {
+            return CompletableFuture.completedFuture(Collections.emptySet());
+        }
+
+        final TypeQuery typeQuery = TypeQueryBuilder
+            .of()
+            .plusPredicates(queryModel -> queryModel.key().isIn(keys))
+            .build();
+
+        return QueryExecutionUtils.queryAll(syncOptions.getCtpClient(), typeQuery)
+                                  .thenApply(types -> types
+                                      .stream()
+                                      .peek(type -> keyToIdCache.put(type.getKey(), type.getId()))
+                                      .collect(toSet()));
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Optional<Type>> fetchType(@Nullable final String key) {
+        if (isBlank(key)) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+
+        return syncOptions.getCtpClient()
+                          .execute(TypeQuery.of().plusPredicates(typeQueryModel -> typeQueryModel.key().is(key)))
+                          .thenApply(PagedResult::head)
+                          .exceptionally(sphereException -> {
+                              syncOptions.applyErrorCallback(format(FETCH_FAILED, key, sphereException),
+                                  sphereException);
+                              return Optional.empty();
+                          });
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Type> createType(@Nonnull final TypeDraft typeDraft) {
+        return syncOptions.getCtpClient().execute(TypeCreateCommand.of(typeDraft));
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Type> updateType(@Nonnull final Type type,
+                                            @Nonnull final List<UpdateAction<Type>> updateActions) {
+        return syncOptions.getCtpClient().execute(TypeUpdateCommand.of(type, updateActions));
```

Comment:
It is very important here to use the `updateResource` of the `BaseService` since it handles the case if we have more than 500 update actions which is not allowed by CTP.

Reply:
Good, it is better now with extending BaseService. ready to review :  https://github.com/commercetools/commercetools-sync-java/pull/333/commits/0c449e0dd18fafee9854b59e39f78818b6972518